packages/agents-hosting/src/oauth/signInContext.ts

```diff
@@ -0,0 +1,390 @@
+   */
+  async getUserToken (): Promise<TokenResponse> {
+    if (!(await this.loadHandler())) {
+      return { token: undefined }
+    }
+    this.logger.info('Getting token from user token service.')
+    return this._authHandler.flow!.getUserToken(this.context)
+  }
+
+  /**
+   * Exchanges the current user token for a new token with specified scopes.
+   * @param scopes - Array of scopes to request for the new token.
+   * @returns A promise that resolves to the exchanged token response.
+   * @remarks
+   * This method checks if the current token is exchangeable (e.g., has audience starting with 'api://')
+   * and performs the appropriate token exchange using MSAL.
+   */
+  async exchangeToken (scopes: string[]): Promise<TokenResponse> {
+    if (!(await this.loadHandler())) {
+      return { token: undefined }
+    }
+    this.logger.info('Exchanging token from user token service.')
+    const tokenResponse = await this._authHandler.flow!.getUserToken(this.context)
+    if (this.isExchangeable(tokenResponse.token)) {
+      return await this.handleObo(tokenResponse.token!, scopes)
+    }
+    return tokenResponse
+  }
+
+  /**
+   * Signs out the user from the current OAuth flow.
+   * @returns {Promise<void>} A promise that resolves when sign out is complete.
+   * @remarks
+   * This method clears the user's token and resets the authentication state.
+   * It also removes the handler state from storage if `useStorageState` is true.
+   */
+  async signOut (): Promise<void> {
+    if (!(await this.loadHandler())) {
+      return
+    }
+
+    this.logger.info('Signing out from the authorization flow.')
+    if (this.isStartedFromRoute) {
+      await this.storage.handler.delete(this.handler.id)
+    }
+    return this._authHandler.flow?.signOut(this.context)
+  }
+
+  /**
+   * Retrieves the token for the current sign-in handler.
+   * @returns {Promise<TokenResponse | undefined>} A promise that resolves to the token response if available, or undefined if not.
+   * @remarks
+   * This method processes the OAuth flow based on the current handler status:
+   * - If the status is 'begin', it starts a new OAuth flow.
+   * - If the status is 'continue', it continues the existing flow.
+   * - If the status is 'success', it retrieves the user token.
+   * - If the status is 'failure', it handles the failure case.
+   */
+  async getToken (): Promise<TokenResponse | undefined> {
+    if (!(await this.loadHandler())) {
+      return { token: undefined }
+    }
+
+    this.logger.debug('Processing authorization flow.')
+    this.logger.debug(`Uses Storage state: ${this.isStartedFromRoute}`)
+    this.logger.debug('Current sign-in state:', this.handler)
+
+    const tokenResponse: TokenResponse | undefined = await {
+      begin: this.begin,
+      continue: this.continue,
+      success: this.success,
+      failure: this.failure
+    }[this.handler.status].bind(this)()
```

Comment:
The use of an object literal to map handler statuses to methods is clever, but it may reduce readability for some maintainers. Consider using a switch statement for clarity.